### PR TITLE
[processing] port shortest path algs to C++

### PR DESCRIPTION
**Pull request by alexbruy**
## Description
Port of the shortest path Processing network analysis algorithms to C++

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: src/analysis/processing/qgsalgorithmserviceareafromlayer.cpp [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+///@cond PRIVATE
+
+QString QgsServiceAreaFromLayerAlgorithm::name() const
+{
+  return QStringLiteral( "serviceareafromlayer" );
+}
+
+QString QgsServiceAreaFromLayerAlgorithm::displayName() const
+{
+  return QObject::tr( "Service area(from layer)" );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmserviceareafromlayer.cpp]: missing space

File: src/analysis/processing/qgsalgorithmserviceareafrompoint.cpp [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+///@cond PRIVATE
+
+QString QgsServiceAreaFromPointAlgorithm::name() const
+{
+  return QStringLiteral( "serviceareafrompoint" );
+}
+
+QString QgsServiceAreaFromPointAlgorithm::displayName() const
+{
+  return QObject::tr( "Service area(from point)" );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmserviceareafrompoint.cpp]: missing space

File: src/analysis/processing/qgsalgorithmserviceareafrompoint.cpp [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+
+  // convert to list so he have same order of points between algorithm runs
+  QList< int > verticesList = vertices.toList();
+  std::sort( verticesList.begin(), verticesList.end() );
+  for ( int v : verticesList )
+  {
+    points.push_back( graph->vertex( v ).point() );
+  }
+  // this produces correct results, but order of point is not the same, so tests failed
+  //QSet< int >::const_iterator it = vertices.constBegin();
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmserviceareafrompoint.cpp]: if you set the 'pk' compare option in the test yaml, does it help with this? That's designed to allow algorithms which export features in random orders.

**Comment by nyalldawson** [comment]
Fanastic stuff! Love it!

File: src/analysis/processing/qgsalgorithmserviceareafrompoint.cpp (hunk 3 above)
Comment (nyalldawson) [reply src/analysis/processing/qgsalgorithmserviceareafrompoint.cpp]: Ouch - that is tricky.

**Comment by alexbruy** [comment]
I decided to split PR, here we have only shortest path algorithms ported, as they pass all tests. I will try to solve issues with service area algs and prepare another PR

**Comment by nyalldawson** [comment]
Good idea - +1 to merge

**Pull request merged**